src/mds/Server.cc

```diff
@@ -1212,25 +1228,55 @@ size_t Server::apply_blacklist(const std::set<entity_addr_t> &blacklist)
   return victims.size();
 }
 
-void Server::journal_close_session(Session *session, int state, Context *on_safe)
+void Server::journal_close_session(Session *session, int state, Context *on_safe, bool on_active)
 {
+  dout(10) << __func__ << " : "
+	   << "("<< on_active << ")"
+	   << session->info.inst
+	   << "(" << session->info.prealloc_inos.size() << "|" << session->pending_prealloc_inos.size() << ")" << dendl;
+
   uint64_t sseq = mds->sessionmap.set_state(session, state);
   version_t pv = mds->sessionmap.mark_projected(session);
   version_t piv = 0;
 
   // release alloc and pending-alloc inos for this session
   // and wipe out session state, in case the session close aborts for some reason
   interval_set<inodeno_t> both;
-  both.insert(session->info.prealloc_inos);
   both.insert(session->pending_prealloc_inos);
+  if (on_active)
+    both.insert(session->info.prealloc_inos);
   if (both.size()) {
     mds->inotable->project_release_ids(both);
     piv = mds->inotable->get_projected_version();
   } else
     piv = 0;
-
-  mdlog->start_submit_entry(new ESession(session->info.inst, false, pv, both, piv),
-			    new C_MDS_session_finish(this, session, sseq, false, pv, both, piv, on_safe));
+  
+  if(!on_active && session->info.prealloc_inos.size()) {
+    dout(10) << "start purge indoes " << session->info.prealloc_inos << dendl;
+    LogSegment* ls = mdlog->get_current_segment();
+    mdlog->start_submit_entry(new ESession(session->info.inst, false, pv, both, piv, session->info.prealloc_inos),
+			      new C_MDS_session_finish(this,
+						       session,
+						       sseq,
+						       false,
+						       pv,
+						       session->pending_prealloc_inos,
+						       piv,
+						       session->info.prealloc_inos,
+						       ls,
+						       on_safe));
+  } else {
+    interval_set<inodeno_t> empty;
+    mdlog->start_submit_entry(new ESession(session->info.inst, false, pv, both, piv, empty),
+			      new C_MDS_session_finish(this,
+						       session,
+						       sseq,
+						       false,
+						       pv,
+						       session->pending_prealloc_inos,
```

Comment:
'both' here ?

Reply:
> 'both' here ?

because the parameter, Context_fin, have default val, so it must be the last parameter. I have to define another function.

Reply:
I mean we should pass 'both' to C_MDS_session_finish()

Reply:
oh.. yes